#7 create link

vadzim-matsiuk opened the pull request:
close #7

Commit:
added entities

added entities for Link, Tag, added links to user

Commit:
added functionality to save and get links

mikita-herasiutsin commented on src/main/java/simplifier/controllers/LinkController.java:
```diff
@@ -0,0 +1,31 @@
+    @PostMapping
+    public ResponseEntity<?> createLink(@RequestBody Link link) {
+        linkService.saveLink(link);
+        return ResponseEntity.status(HttpStatus.CREATED).build();
```
Return saved link to let know how shortened link looks like

mikita-herasiutsin commented on src/main/java/simplifier/controllers/LinkController.java:
```diff
@@ -0,0 +1,31 @@
+    }
+
+    @PostMapping
+    public ResponseEntity<?> createLink(@RequestBody Link link) {
```
Make dto object without fields that should not be saved, e.g. id and clicks. API should be easy to use by clients. Change incoming json to this structure
{
	"originalLink": "https://github.com/matsiukvadzim/URL-Simplifier/pull/10/files",
	"shortenedLink": "short",
	"description": "PR",
	"tags": [
               "quality",
               "test"
        ]
	"author":"putin"
}

mikita-herasiutsin commented on src/main/java/simplifier/services/LinkServiceImpl.java:
```diff
@@ -0,0 +1,55 @@
+        User existingUser = userService.findByUsername(link.getAuthor().getUsername());
+        link.setAuthor(existingUser);
+        linkRepository.save(link);
+        link.setShortenedLink(simplifyService.encode(link.getId()));
```
Shortened url would be overridden even if it was listed

mikita-herasiutsin commented on src/main/java/simplifier/services/TagServiceImpl.java:
```diff
@@ -0,0 +1,27 @@
+
+    @Override
+    public List<Tag> saveAllTags(List<Tag> tags) {
+        return StreamSupport.stream(tagRepository.saveAll(tags).spliterator(),false)
```
Just return saved tags

mikita-herasiutsin commented on src/main/resources/application.properties:
```diff
@@ -1,4 +1,4 @@
-spring.datasource.url=jdbc:postgresql://ec2-54-221-207-184.compute-1.amazonaws.com:5432/d4gro9nv6bonah?user=ojucslvrmhfwdn&password=e1f0db18b4b55bd520ebedc89d8ebffcea06c9fc5138b2719f9c051508f87a52&sslmode=require
-spring.datasource.username=ojucslvrmhfwdn
```
Why did you comment that?

mikita-herasiutsin commented on src/main/java/simplifier/services/LinkServiceImpl.java:
```diff
@@ -0,0 +1,55 @@
+import simplifier.repositories.LinkRepository;
+
+@Service
+public class LinkServiceImpl implements LinkService {
```
Add tests

mikita-herasiutsin commented on src/main/java/simplifier/controllers/LinkController.java:
```diff
@@ -0,0 +1,31 @@
+
+@RestController
+@RequestMapping("/link")
+public class LinkController {
```
Add tests

mikita-herasiutsin commented on src/main/java/simplifier/services/LinkSimplifyServiceImpl.java:
```diff
@@ -0,0 +1,29 @@
+import org.springframework.stereotype.Service;
+
+@Service
+public class LinkSimplifyServiceImpl implements LinkSimplifyService {
```
Add tests

mikita-herasiutsin commented on src/main/java/simplifier/services/TagServiceImpl.java:
```diff
@@ -0,0 +1,27 @@
+import java.util.stream.StreamSupport;
+
+@Service
+public class TagServiceImpl implements TagService {
```
Add tests

mikita-herasiutsin commented on src/main/java/simplifier/model/Link.java:
```diff
@@ -0,0 +1,97 @@
+
+    private String originalLink;
+
+    private String shortenedLink;
```
should be unique

mikita-herasiutsin commented on src/main/java/simplifier/model/Tag.java:
```diff
@@ -0,0 +1,45 @@
+    @GeneratedValue(strategy = GenerationType.AUTO)
+    private Integer id;
+
+    private String name;
```
should be unique

Commit:
Tests were added. Several methods were refactored

Tests were added for following services:
- LinkServiceImpl unit test
- LinkSimplifyServiceImpl unit test
- TagServiceImpl unit test
- LinkController integration test

Now it's possible to create link with custom shortened .
Save method in tagService was refactored, so it's can create or update tags.
Also, custom json deserializer was added.

mikita-herasiutsin commented on src/main/java/simplifier/services/LinkServiceImpl.java:
```diff
@@ -0,0 +1,66 @@
+        link.setTags(tagService.saveOrUpdateTags(link));
+        User existingUser = userRepository.findByUsername(link.getAuthor().getUsername());
+        link.setAuthor(existingUser);
+        if (link.getShortenedLink() == null || !checkIsShortenedUnique(link)) {
```
Link's shortened URL will be auto generated if it is not unique. This is an unexpected behaviour. It would be better to return message that URL is not unique

mikita-herasiutsin commented on src/test/java/simplifier/controllers/LinkControllerIT.java:
```diff
@@ -0,0 +1,130 @@
+    }
+
+    @Before
+    public void setUp() throws Exception {
```
Use test data from files

mikita-herasiutsin commented on src/main/java/simplifier/services/LinkServiceImpl.java:
```diff
@@ -0,0 +1,66 @@
+    @Override
+    public Link saveLink(Link link) {
+        link.setTags(tagService.saveOrUpdateTags(link));
+        User existingUser = userRepository.findByUsername(link.getAuthor().getUsername());
```
User may be not found

mikita-herasiutsin commented on src/main/java/simplifier/components/LinkJsonComponent.java:
```diff
@@ -0,0 +1,76 @@
+import java.util.List;
+
+@JsonComponent
+public class LinkJsonComponent {
```
Remove this component and use dto objects with transformers instead

mikita-herasiutsin commented on src/main/java/simplifier/repositories/TagRepository.java:
```diff
@@ -0,0 +1,9 @@
+
+public interface TagRepository extends CrudRepository<Tag, Integer> {
+
+    Tag findByName(String name);
```
Change return type to Optional<Tag> with changing if/else to map/orElse

mikita-herasiutsin commented on src/main/java/simplifier/services/UserService.java:
```diff
@@ -7,4 +7,6 @@
 
     Optional<User> saveUser(User user);
+
+    User findByUsername(String username);
```
Change return type to Optional<User> with changing if/else to map/orElse

Commit:
DTOs and validator were added

mikita-herasiutsin commented on src/main/java/simplifier/controllers/LinkController.java:
```diff
@@ -1,31 +1,43 @@
     @PostMapping
-    public ResponseEntity<?> createLink(@RequestBody Link link) {
+    public ResponseEntity<?> createLink(@Valid @RequestBody LinkCreationDto linkDto) {
+        Link link = linkMapper.linkDtoToLink(linkDto);
```
This should be moved to service

mikita-herasiutsin commented on src/main/java/simplifier/validators/LinkAuthorValidator.java:
```diff
@@ -0,0 +1,24 @@
+
+    @Override
+    public boolean isValid(String value, ConstraintValidatorContext context) {
+        return userRepository.findByUsername(value)
```
may be simplified to isEmpty()

mikita-herasiutsin commented on src/main/java/simplifier/validators/LinkShortenedValidator.java:
```diff
@@ -0,0 +1,24 @@
+
+    @Override
+    public boolean isValid(String value, ConstraintValidatorContext context) {
+        return linkRepository.findByShortenedLink(value)
```
may be simplified to isEmpty()

Commit:
UserService was refactored. Added tests.

mikita-herasiutsin commented on src/main/java/simplifier/services/TagServiceImpl.java:
```diff
@@ -0,0 +1,41 @@
+        List<Tag> tags = new ArrayList<>();
+
+        for (Tag tag : link.getTags()) {
+            Tag currentTag = tagRepository.findByName(tag.getName());
```
Replace findByName() to method which accepts list of tag's names to avoid multiple calls to db in for()
FYI https://stackoverflow.com/questions/18987292/spring-crudrepository-findbyinventoryidslistlong-inventoryidlist-equivalen

mikita-herasiutsin commented on src/main/java/simplifier/services/TagServiceImpl.java:
```diff
@@ -0,0 +1,41 @@
+                tags.add(currentTag);
+            } else {
+                tag.addLink(link);
+                Tag savedTag = tagRepository.save(tag);
```
Move this outside of for() and save all tags at once

mikita-herasiutsin commented on src/main/java/simplifier/services/UserService.java:
```diff
@@ -7,4 +7,6 @@
 
     Optional<User> saveUser(User user);
+
+    Optional<User> findByUsername(String username);
```
Method is nod used

mikita-herasiutsin commented on src/test/java/simplifier/services/TagServiceImplTest.java:
```diff
@@ -0,0 +1,75 @@
+
+    @Test
+    public void saveNewTag() {
+        Link link = new Link();
```
may be moved to private method

mikita-herasiutsin commented on src/test/java/simplifier/services/LinkServiceImplTest.java:
```diff
@@ -0,0 +1,139 @@
+    private LinkGetterDto getterDto = new LinkGetterDto();
+
+    @Before
+    public void setUp() throws Exception {
```
redundant throws declaration

mikita-herasiutsin commented on src/test/resources/LinkWithShortened.JSON:
```diff
@@ -0,0 +1,10 @@
+{
```
rename to just "validLink.json"

Commit:
TagService was fixed

Equals & hashcode were overridden.

mikita-herasiutsin commented on src/test/java/simplifier/services/LinkSimplifyServiceImplTest.java:
```diff
@@ -0,0 +1,23 @@
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.is;
+
+public class LinkSimplifyServiceImplTest {
```
Add test for decode() method

mikita-herasiutsin replied:
Skip this comment. It was a mistake

Commit:
Added test to TagServiceImplTest
